### PR TITLE
Send LSP message for early exits on errors

### PR DESCRIPTION
**Pull request by soam-stripe**
Sends an LSP message (`window/showMessage`) before exiting due to errors.

### Motivation

#3570 

### Test plan

See included automated tests.

### CODE REVIEW

File: main/lsp/protocol.cc [hunk 1]
```diff
@@ -140,12 +146,18 @@ optional<unique_ptr<core::GlobalState>> LSPLoop::runLSP(shared_ptr<LSPInput> inp
...
-            [&messageQueue, &messageQueueMutex, logger = logger](int watchmanExitCode) {
+            [&messageQueue, &messageQueueMutex, logger = logger, config = this->config](int watchmanExitCode,
+                                                                                        string const &msg) {
                 {
                     absl::MutexLock lck(&messageQueueMutex);
                     if (!messageQueue.terminate) {
                         messageQueue.terminate = true;
                         messageQueue.errorCode = watchmanExitCode;
+                        if (watchmanExitCode != 0) {
+                            auto params = make_unique<ShowMessageParams>(MessageType(1 /* error */), msg);
```
Comment (jez-stripe) [inline main/lsp/protocol.cc:157]:
```suggestion
                            auto params = make_unique<ShowMessageParams>(MessageType::Error, msg);
```

you can use the enum name here (and all other places)

File: main/lsp/protocol.cc [hunk 2]
```diff
@@ -118,8 +120,12 @@ optional<unique_ptr<core::GlobalState>> LSPLoop::runLSP(shared_ptr<LSPInput> inp
     auto &logger = config->logger;
     if (!opts.disableWatchman) {
         if (opts.rawInputDirNames.size() != 1 || !opts.rawInputFileNames.empty()) {
-            logger->error("Watchman support currently only works when Sorbet is run with a single input directory. If "
-                          "Watchman is not needed, run Sorbet with `--disable-watchman`.");
```
Comment (jez-stripe) [inline main/lsp/protocol.cc:122]: I wonder if we want to keep the error in the log as well, just in case.

**Comment by jez-stripe** [comment]
one thing i'm curious about: can you try this in VS Code and take a screenshot of what the UI looks like?

I'm curious how the UI for this vs a `window/showMessageRequest` *Request* works (`window/showMessage` is a notification)

File: main/lsp/protocol.cc (hunk 2 above)
Comment (soam-stripe) [reply main/lsp/protocol.cc:122]: It is still logged, see line 125 below.

**Comment by jvilk-stripe** [comment]
> I'm curious how the UI for this vs a window/showMessageRequest Request works (window/showMessage is a notification)

`showMessageRequest` is only relevant if we're actually presenting the user with some choice. I don't think that's the case here.

Also, the LSP server currently doesn't have support for sending requests to the client; it would require some work to add that in.

**Comment by jez** [comment]
> > I'm curious how the UI for this vs a window/showMessageRequest Request works (window/showMessage is a notification)
> 
> `showMessageRequest` is only relevant if we're actually presenting the user with some choice. I don't think that's the case here.
> 
> Also, the LSP server currently doesn't have support for sending requests to the client; it would require some work to add that in.

Ah that's too bad.

I was thinking that maybe we could get them to click a "Confirm" prompt or something—I noticed when I was testing this in Neovim that when I prototyped something hacky with `window/showMessage`, the UI real estate for the message is a single terminal line at the bottom of Neovim, which other plugins and Neovim itself sometimes write messages too.

In particular, since these messages came right before the plugin crashed, my language client plugin for neovim was sending it's own `[LC] Language server quit unexpectedly` message, clobbering the helpful message that Sorbet just sent.

In Neovim the entire history of messages are available if you type `:messages` so it's still there, but I was curious to see what might have happened in the UI if it was a `window/showMessageRequest` (because my language client has a different UI for presenting a selection of options).

But if there's no prior art for messages from Sorbet to client yet, don't worry about it—the `window/showMessage` notification is better than only `logger->error` messages.

But I am still curious to see what VS Code's UI looks like by comparison!

**Comment by soam-stripe** [comment]
Screen Shot 2020-10-28 at 11 56 15 AM

Screen Shot 2020-10-28 at 11 57 29 AM

Here's how the message shows up. The second one is after clicking message and then the down arrow on it to expand. Unfortunately you have read the expanded part of the message within about 5 seconds, because vscode gets into a retry loop -- so the message keeps getting reissued in the collapsed form. Bleh.

(You can click on the sorbet extension name and click view output, to see the output of the logger.)

**Comment by jez** [comment]
This looks good to me but i'll defer to @jvilk-stripe

**Review by jvilk-stripe — APPROVED** [review]
Thanks for testing!

**Pull request merged**